patches/server/0981-Rewrite-the-Amethyst-logic-to-throw-both-BlockSpread.patch

```diff
@@ -0,0 +1,48 @@
+
+diff --git a/src/main/java/net/minecraft/world/level/block/BuddingAmethystBlock.java b/src/main/java/net/minecraft/world/level/block/BuddingAmethystBlock.java
+index bedccb8717d08d5a60058445b04ddff149e7d36c..1cf30898acd8c908cf72e08eae8d8f040486fb3b 100644
+--- a/src/main/java/net/minecraft/world/level/block/BuddingAmethystBlock.java
++++ b/src/main/java/net/minecraft/world/level/block/BuddingAmethystBlock.java
+@@ -29,11 +29,18 @@ public class BuddingAmethystBlock extends AmethystBlock {
+             Direction enumdirection = BuddingAmethystBlock.DIRECTIONS[random.nextInt(BuddingAmethystBlock.DIRECTIONS.length)];
+             BlockPos blockposition1 = pos.relative(enumdirection);
+             BlockState iblockdata1 = world.getBlockState(blockposition1);
++            BlockState iblockdata2 = null; //Paper - move here for reuse
+             Block block = null;
+ 
+             if (BuddingAmethystBlock.canClusterGrowAtState(iblockdata1)) {
+                 block = Blocks.SMALL_AMETHYST_BUD;
+-            } else if (iblockdata1.is(Blocks.SMALL_AMETHYST_BUD) && iblockdata1.getValue(AmethystClusterBlock.FACING) == enumdirection) {
++                // Paper start - BlockSpreadEvent
++                iblockdata2 = (BlockState) ((BlockState) block.defaultBlockState().setValue(AmethystClusterBlock.FACING, enumdirection)).setValue(AmethystClusterBlock.WATERLOGGED, iblockdata1.getFluidState().getType() == Fluids.WATER);
++                org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(world, pos, blockposition1, iblockdata2);
++                return;
++            } // Paper end
++
++            if (iblockdata1.is(Blocks.SMALL_AMETHYST_BUD) && iblockdata1.getValue(AmethystClusterBlock.FACING) == enumdirection) { // Paper - remove else
+                 block = Blocks.MEDIUM_AMETHYST_BUD;
+             } else if (iblockdata1.is(Blocks.MEDIUM_AMETHYST_BUD) && iblockdata1.getValue(AmethystClusterBlock.FACING) == enumdirection) {
+                 block = Blocks.LARGE_AMETHYST_BUD;
+@@ -42,9 +49,9 @@ public class BuddingAmethystBlock extends AmethystBlock {
+             }
+ 
+             if (block != null) {
+-                BlockState iblockdata2 = (BlockState) ((BlockState) block.defaultBlockState().setValue(AmethystClusterBlock.FACING, enumdirection)).setValue(AmethystClusterBlock.WATERLOGGED, iblockdata1.getFluidState().getType() == Fluids.WATER);
++                iblockdata2 = (BlockState) ((BlockState) block.defaultBlockState().setValue(AmethystClusterBlock.FACING, enumdirection)).setValue(AmethystClusterBlock.WATERLOGGED, iblockdata1.getFluidState().getType() == Fluids.WATER);
+ 
+-                org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(world, pos, blockposition1, iblockdata2); // CraftBukkit
++                org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(world, blockposition1, iblockdata2); //Paper - switch to BlockGrowEvent
```

Comment:
This whole logic could be done here unless i miss something with:
if block is small bud: call the spread event
otherwise: call the grow event

That way the diff is smaller, also don't forget the space between //  and Paper and no need to
write the version of this patch per commit

Reply:
I tried a few different ways to make the patch smaller, but I couldn't find a way to also make it easy to see what's going on as well.

Originally I rewrote it as a switch statement to make the if/elseif/else loop a bit more optimized, but ran into issues understanding _what_ was being done, so I rewrote it with the better local names to understand things (Version 1, it's also something I picked up with the code I write for my other projects and stuff for my job so it's more of a habit to version my patches). Which was denied because, well, big diff. So I undid the local name changes and fixed up some of the mistakes from that (Version 2 and 3). Which is 10 line changes.

To try and put it in the `if (block != null)` block, I would need to either but in a local boolean to track if the shard spread (which would be 2 line changes) or another if check inside that if block (like `if (block == Block.SMALL_AMETHYST_BUD)`, which would require indenting everything else inside the first if block, making all those line changes as well).

Suggestion 1 (using spigot's version because I have that open right now)
``` java
if (randomsource.nextInt(5) == 0) {
    EnumDirection enumdirection = BuddingAmethystBlock.DIRECTIONS[randomsource.nextInt(BuddingAmethystBlock.DIRECTIONS.length)];
    BlockPosition blockposition1 = blockposition.relative(enumdirection);
    IBlockData iblockdata1 = worldserver.getBlockState(blockposition1);
    Block block = null;
    boolean spread = false; //1

    if (canClusterGrowAtState(iblockdata1)) {
        block = Blocks.SMALL_AMETHYST_BUD;
        spread = true; //2
    } else if (iblockdata1.is(Blocks.SMALL_AMETHYST_BUD) && iblockdata1.getValue(AmethystClusterBlock.FACING) == enumdirection) {
        block = Blocks.MEDIUM_AMETHYST_BUD;
    } else if (iblockdata1.is(Blocks.MEDIUM_AMETHYST_BUD) && iblockdata1.getValue(AmethystClusterBlock.FACING) == enumdirection) {
        block = Blocks.LARGE_AMETHYST_BUD;
    } else if (iblockdata1.is(Blocks.LARGE_AMETHYST_BUD) && iblockdata1.getValue(AmethystClusterBlock.FACING) == enumdirection) {
        block = Blocks.AMETHYST_CLUSTER;
    }

    if (block != null) {
        IBlockData iblockdata2 = (IBlockData) ((IBlockData) block.defaultBlockState().setValue(AmethystClusterBlock.FACING, enumdirection)).setValue(AmethystClusterBlock.WATERLOGGED, iblockdata1.getFluidState().getType() == FluidTypes.WATER);
        if (spread) { //3
            org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(worldserver, blockposition, blockposition1, iblockdata2); //4
        } else { //5
            org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(worldserver, blockposition1, iblockdata2); //6
        } //7
    }

}
```

Suggestion 2 (also using spigot's version)
``` java
if (randomsource.nextInt(5) == 0) {
    EnumDirection enumdirection = BuddingAmethystBlock.DIRECTIONS[randomsource.nextInt(BuddingAmethystBlock.DIRECTIONS.length)];
    BlockPosition blockposition1 = blockposition.relative(enumdirection);
    IBlockData iblockdata1 = worldserver.getBlockState(blockposition1);
    Block block = null;

    if (canClusterGrowAtState(iblockdata1)) {
        block = Blocks.SMALL_AMETHYST_BUD;
    } else if (iblockdata1.is(Blocks.SMALL_AMETHYST_BUD) && iblockdata1.getValue(AmethystClusterBlock.FACING) == enumdirection) {
        block = Blocks.MEDIUM_AMETHYST_BUD;
    } else if (iblockdata1.is(Blocks.MEDIUM_AMETHYST_BUD) && iblockdata1.getValue(AmethystClusterBlock.FACING) == enumdirection) {
        block = Blocks.LARGE_AMETHYST_BUD;
    } else if (iblockdata1.is(Blocks.LARGE_AMETHYST_BUD) && iblockdata1.getValue(AmethystClusterBlock.FACING) == enumdirection) {
        block = Blocks.AMETHYST_CLUSTER;
    }

    if (block != null) {
        IBlockData iblockdata2 = (IBlockData) ((IBlockData) block.defaultBlockState().setValue(AmethystClusterBlock.FACING, enumdirection)).setValue(AmethystClusterBlock.WATERLOGGED, iblockdata1.getFluidState().getType() == FluidTypes.WATER);
        if(block == Blocks.SMALL_AMETHYST_BUD) { //1
            org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(worldserver, blockposition, blockposition1, iblockdata2); //2
        } else { //3
            org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(worldserver, blockposition1, iblockdata2); //4
        } //5
    }

}
```

Suggestion 2 is the smallest I could make it. Between that and mine, Suggestion 2 still has to go through that entire if/elseif/else block whereas mine returns after the spread event, skipping the rest of the checks that don't need to be checked. It is a negligible savings in processing, but a savings nonetheless, and trying to optimize the logic was what I was aiming for when working on this.